src/Tests/dotnet-MsiInstallation.Tests/VMControl.cs

```diff
@@ -0,0 +1,387 @@
+
+
+
+            if (!string.IsNullOrEmpty(snapshotName))
+            {
+                await RenameSnapshotAsync(newSnapshot.id, snapshotName);
+            }
+
+            Log.WriteLine($"Created snapshot {snapshotName} ({newSnapshot.id})");
+
+            return newSnapshot.id;
+        }
+
+        public async Task ApplySnapshotAsync(string snapshotId)
+        {
+            if (!await ApplySnapshotIfExistsAsync(snapshotId))
+            {
+                throw new Exception("Snapshot not found: " + snapshotId);
+            }
+        }
+
+        public async Task<bool> ApplySnapshotIfExistsAsync(string snapshotId)
+        {
+            var snapshots = GetSnapshotInstances();
+
+            var snapshot = snapshots.SingleOrDefault(s => s.CimInstanceProperties["ConfigurationID"].Value.ToString() == snapshotId);
+
+            if (snapshot == null)
+            {
+                return false;
+            }
+
+            await SetRunningAsync(false);
+
+            var snapshotService = _session.EnumerateInstances(virtNamespace, "Msvm_VirtualSystemSnapshotService").First();
+
+            CimMethodParametersCollection cimMethodParameters = new CimMethodParametersCollection
+            {
+                CimMethodParameter.Create("Snapshot", snapshot, CimType.Reference, CimFlags.In),
+            };
+
+            Log.WriteLine($"Applying snapshot {snapshot.CimInstanceProperties["ElementName"].Value} ({snapshot.CimInstanceProperties["ConfigurationID"].Value})");
+
+            var result = _session.InvokeMethod(virtNamespace, snapshotService, "ApplySnapshot", cimMethodParameters);
+
+            await WaitForJobSuccess(result);
+
+            await SetRunningAsync(true);
+
+            return true;
+        }
+
+        public async Task SetRunningAsync(bool running)
+        {
+            VMEnabledState getCurrentState()
+            {
+                var state = (VMEnabledState) (UInt16)VMInstance.CimInstanceProperties["EnabledState"].Value;
+                return state;
+            }
+
+            var targetState = running ? VMEnabledState.Enabled : VMEnabledState.Disabled;
+
+            if (getCurrentState() == targetState)
+            {
+                return;
+            }
+
+            Log.WriteLine($"Changing EnabledState from {getCurrentState()} to {targetState}");
+
+            var methodParameters = new CimMethodParametersCollection()
+            {
+                CimMethodParameter.Create("RequestedState", (UInt16) targetState, CimFlags.In)
+            };
+
+
+            var result = _session.InvokeMethod(virtNamespace, VMInstance, "RequestStateChange", methodParameters);
+
+            await WaitForJobSuccess(result);
+
+            for (int i = 0; i < 10 && getCurrentState() != targetState; i++)
+            {
+                Log.WriteLine("Waiting for state change...");
+                await Task.Delay(250);
+            }
```

Comment:
Maybe a warning would be better.

Reply:
I don't think we have a way to generate a warning here.  If it doesn't end up syncing to the right state the next operation will probably fail.  I'm hesitant to make changes to this now because it seems to work well as is but took a bunch of trial and error to get to that point.